### PR TITLE
feat: add documentation generation to CI

### PR DESCRIPTION
**Pull request by bollu**
This adds `doc-gen4` adds a dependency, a shell script to generate docs, and a github action to run doc generation on CI. This ensures that we are always in a state where we can successfully generate the docs.

### CODE REVIEW

**Comment by bollu** [comment]
@shigoel I've disabled uploading the docs. We still generate the docs, to make sure that we're always in a state that we can perform documentation generation.

File: Lib.lean [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/-
+Copyright (c) 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Siddharth Bhat
+
+We collect all modules in LNsym to build a single library for documentation generation.
```
Comment (shigoel) [inline Lib.lean:6]:
```suggestion
We collect all modules in LNSym to build a single library for documentation generation.
```

File: Lib.lean [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/-
+Copyright (c) 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (shigoel) [inline Lib.lean:2]:
```suggestion
Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: .github/workflows/doc.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  pages: write
+  id-token: write
+
+jobs:
+  build:
+    name: build and deploy documentation.
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout 🛎️
+        uses: actions/checkout@v3
```
Comment (shigoel) [inline .github/workflows/doc.yml:19]:
```suggestion
        uses: actions/checkout@v4
```

as per `makefile.yml`

File: README.md [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # LNSym: Native Code Symbolic Simulator in Lean
-[![Makefile CI](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml/badge.svg)](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml)
+[![Makefile CI](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml/badge.svg)](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml) [Docs](https://leanprover.github.io/LNSym/)
```
Comment (shigoel) [inline README.md:2]:
```suggestion
[![Makefile CI](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml/badge.svg)](https://github.com/leanprover/LNSym/actions/workflows/makefile.yml)
```

File: docgen.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+set -o xtrace
+set -e
+lake build Lib:docs
```
Comment (shigoel) [inline docgen.sh:3]:
```suggestion
lake build Doc:docs
```

File: Lib.lean
Comment (shigoel) [inline Lib.lean:1]: TODO: rename `Lib` to `Doc` to communicate intent.

File: .github/workflows/doc.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+      - name: Install elan 🕑
+        run: |
+          set -o pipefail
+          curl https://raw.githubusercontent.com/leanprover/elan/master/elan-init.sh -sSf | sh -s -- --default-toolchain none -y
+          ~/.elan/bin/lean --version
+          echo "$HOME/.elan/bin" >> $GITHUB_PATH
```
Comment (shigoel) [inline .github/workflows/doc.yml:26]: Refactor to use `leanprover/lean-action`.

**Review by shigoel — COMMENTED** [review]
LGTM modulo comments.

**Comment by bollu** [comment]
@shigoel I've moved to `Doc.lean`, and updated the CI script to use the lean action.

**Review by shigoel — APPROVED** [review]
Thanks!

**Pull request merged**